### PR TITLE
Generate basic statistics

### PR DESCRIPTION
**Pull request by sargon**
Generate basic statistics so they could be collected in time series database.

### CODE REVIEW

File: ddhcp.c [hunk 1]
```diff
@@ -233,9 +239,11 @@ void ddhcp_dhcp_renewlease(struct ddhcp_mcast_packet* packet, ddhcp_config* conf
...
     DEBUG("ddhcp_dhcp_renewlease(...): %i ACK\n", ret);
     answer = new_ddhcp_packet(DDHCP_MSG_LEASEACK, config);
+    statistics_record(config, STAT_DIRECT_SEND_LEASEACK, 1);
   } else if (ret == 1) {
     DEBUG("ddhcp_dhcp_renewlease(...): %i NAK\n", ret);
     answer = new_ddhcp_packet(DDHCP_MSG_LEASENAK, config);
+    statistics_record(config, STAT_DIRECT_SEND_LEASENAK, 1);
     // TODO Can we hand over the block?
   } else {
     // Unexpected behaviour
```
Comment (BenBE) [inline ddhcp.c]: Shouldn't we record statistics on this branch, even if it shouldn't be called normally?

File: ddhcpctl.c [hunk 2]
```diff
@@ -33,7 +33,7 @@ int main(int argc, char** argv) {
     exit(1);
   }
 
-  while ((c = getopt(argc, argv, "C:t:l:bdho:r:v:V")) != -1) {
+  while ((c = getopt(argc, argv, "C:t:l:bdho:sSr:v:V")) != -1) {
```
Comment (BenBE) [inline ddhcpctl.c]: `...r:sSv:V` instead of `...sSr:v:V`

File: ddhcpctl.c [hunk 3]
```diff
@@ -51,6 +51,16 @@ int main(int argc, char** argv) {
       buffer[0] = (uint8_t) DDHCPCTL_DHCP_OPTIONS_SHOW;
       break;
 
+    case 's':
```
Comment (BenBE) [inline ddhcpctl.c]: Options handled even if `DDHCPD_STATISTICS` disabled.

File: dhcp.c [hunk 4]
```diff
@@ -144,6 +145,7 @@ int dhcp_process(uint8_t* buffer, ssize_t len, ddhcp_config* config) {
 
     switch (message_type) {
     case DHCPDISCOVER:
+      statistics_record(config, STAT_DHCP_RECV_DISCOVER, 1);
       ret = dhcp_hdl_discover(config->client_socket, &dhcp_packet_buf, config);
 
       if (ret == 1) {
```
Comment (BenBE) [inline dhcp.c]: Maybe a probe here for new blocks inquired after received discover vs. new blocks due to other conditions?

File: ddhcp.c [hunk 5]
```diff
@@ -297,7 +309,8 @@ void ddhcp_dhcp_leasenak(struct ddhcp_mcast_packet* request, ddhcp_config* confi
     DEBUG("ddhcp_dhcp_leaseack(...): No matching packet found, message ignored\n");
   } else {
     // Process packet
-    dhcp_nack(config->client_socket, packet);
+    statistics_record(config, STAT_DHCP_SEND_PKG, 1);
```
Comment (BenBE) [inline ddhcp.c]: Missing second stat on what this packet was sent for as found at other locations.

File: dhcp.c [hunk 6]
```diff
@@ -273,6 +281,7 @@ int dhcp_rhdl_ack(int socket, struct dhcp_packet* request, ddhcp_config* config)
     return 1;
   }
 
+  statistics_record(config, STAT_DHCP_SEND_PKG, 1);
```
Comment (BenBE) [inline dhcp.c]: Should indicate what this packet was sent for.

File: dhcp.c [hunk 7]
```diff
@@ -350,7 +359,12 @@ int dhcp_hdl_request(int socket, struct dhcp_packet* request, ddhcp_config* conf
         // TODO Error handling
         dhcp_packet_list_add(&config->dhcp_packet_cache, request);
 
-        send_packet_direct(packet, &lease_block->owner_address, config->server_socket, config->mcast_scope_id);
+        statistics_record(config, STAT_DIRECT_SEND_PKG, 1);
+        statistics_record(config, STAT_DIRECT_SEND_RENEWLEASE, 1);
+        ssize_t bytes_send = send_packet_direct(packet, &lease_block->owner_address, config->server_socket, config->mcast_scope_id);
+        statistics_record(config, STAT_DIRECT_SEND_BYTE, (long int) bytes_send);
+        UNUSED(bytes_send);
+        UNUSED(config);
```
Comment (BenBE) [inline dhcp.c]: `config` is used in line 360 AFAIC. No need to mark it `UNUSED` here.

File: dhcp.c [hunk 8]
```diff
@@ -471,7 +485,12 @@ int dhcp_nack(int socket, dhcp_packet* from_client) {
     DHCPNAK
   });
 
-  dhcp_packet_send(socket, packet);
+  statistics_record(config, STAT_DHCP_SEND_NAK, 1);
```
Comment (BenBE) [inline dhcp.c]: Missing marker for number of packets sent:
```c
statistics_record(config, STAT_DHCP_SEND_PKG, 1);
```

File: dhcp.c [hunk 9]
```diff
@@ -504,7 +523,10 @@ int dhcp_ack(int socket, dhcp_packet* request, ddhcp_block* lease_block, uint32_
   }
 
   DEBUG("dhcp_ack(...): offering address %i %s\n", lease_index, inet_ntoa(packet->yiaddr));
-  dhcp_packet_send(socket, packet);
+  statistics_record(config, STAT_DHCP_SEND_ACK, 1);
```
Comment (BenBE) [inline dhcp.c]: Missing marker for number of packets sent:
```c
statistics_record(config, STAT_DHCP_SEND_PKG, 1);
```

File: types.h [hunk 10]
```diff
@@ -15,6 +15,40 @@ typedef uint8_t ddhcp_node_id[8];
 #define NODE_ID_CLEAR(id) memset(id,'\0',sizeof(ddhcp_node_id))
 #define NODE_ID_CP(dest,src) memcpy(dest,src,sizeof(ddhcp_node_id))
 
+// statistic fields
```
Comment (BenBE) [inline types.h]: I'd prefer to use an enum here.

File: ddhcpctl.c (hunk 2 above)
Comment (sargon) [reply ddhcpctl.c]: What is your point here?

File: dhcp.c (hunk 4 above)
Comment (sargon) [reply dhcp.c]: There is no direct connection to inquires and discoveries. The inquire process is triggered only by the spare block ensuring mechanism.

File: ddhcpctl.c (hunk 2 above)
Comment (BenBE) [reply ddhcpctl.c]: Order of options in the string; suggestion was to have it alphabetical.

Comment (sargon) [reply ddhcpctl.c]: K fixed, but for the whole string.

**Pull request merged**